File: python/freetensor/core/auto_schedule.py
import freetensor_ffi as ffi
import xgboost as xgb
import numpy as np
import os


class AutoSchedule(ffi.AutoSchedule):

    def __init__(self,
                 schedule,
                 target,
                 device,
                 *,
                 population=64,
                 explore_ratio=0.1,
                 tag="",
                 min_block_size=0,
                 continue_training=False,
                 random_seed=None,
                 rule_set=None,
                 verbose=0,
                 lower_func_submit_api=None):
        '''
        Automatic scheduler

        Parameters
        ----------
        schedule : Schedule
            A Schedule object to apply schedules onto
        target : Target
            The type of devices to compile to
        population : int
            How many programs to test in each iteration
        explore_ratio : float
            Portion of random programs in the population. Higher ratio focuses on
            exploration, while lower ratio focuses on exploitation
        continue_trianing : bool
            Continue to train an existing XGBoost model file if found
        random_seed : Optional[int]
            Random seed. Setting a deterministic random seed and using a fixed OpenMP
            thread count (since we are using thread-local random number generators)
            resulting deterministic pseudo random numbers, but please note that the
            whole auto-scheduling procedure is still non-deterministic, becuase it
            measures real performance. Default to a non-deterministic seed
        rule_set : Optional[set]
            Explicitly control over what rules to use. None for defualt rules
        verbose : int
            Verbosity level. 0 = print nothing, 1 = print tuning progress, 2 = print
            extra info mation of each rule
        '''

        self.population = population
        self.n_explore = int(population * explore_ratio)
        self.n_exploit = population - self.n_explore
        self.model = None
        self.xgb_params = {}
        self.save_file_name = tag + "_xgb.model"
        if continue_training and os.path.isfile(self.save_file_name):
            self.model = xgb.Booster()
            self.model.load_model(self.save_file_name)
        self.verbose = verbose
        self.lower_func_submit_api = lower_func_submit_api

        def predict_func(features):
            return self.predict(features)

        def update_func(features, times):
            return self.update(features, times)

        super(AutoSchedule,
              self).__init__(schedule, target, device, predict_func,
                             update_func, tag, min_block_size, random_seed,
                             rule_set, verbose, lower_func_submit_api)

    def set_params(self, *args, **kws):
        super(AutoSchedule, self).set_params(args, kws)

    def run(self, iteration):
        for i in range(iteration):
            if self.verbose >= 1:
                print("Iteration", i)
            self.search_one_round(self.population, self.n_exploit,
                                  self.n_explore)
        return self.get_best_schedule()

    def predict(self, features):
        if not self.model:
            return [1] * len(features)
        return self.model.predict(xgb.DMatrix(np.array(features), missing=-1))

    def update(self, features, times):
        dtrain = xgb.DMatrix(np.array(features), np.array(times), missing=-1)
        self.model = xgb.train(self.xgb_params, dtrain, xgb_model=self.model)
        self.model.save_model(self.save_file_name)

File: ffi/auto_schedule.cc
#include <auto_schedule/auto_schedule.h>
#include <driver/array.h>
#include <ffi.h>
#include <schedule.h>

namespace freetensor {

using namespace pybind11::literals;

void init_ffi_auto_schedule(py::module_ &m) {
    py::class_<Sketch>(m, "Sketch")
        .def("get_annotation", &Sketch::getAnnotation);
    py::class_<AutoSchedule>(m, "AutoSchedule")
        .def(py::init<
                 const Schedule &, const Ref<Target> &, const Ref<Device> &,
                 const std::function<AutoSchedule::Predicts(
                     const AutoSchedule::Features &)> &,
                 const std::function<void(const AutoSchedule::Features &,
                                          const AutoSchedule::Predicts &)> &,
                 std::string, int, std::optional<size_t>,
                 const std::optional<std::unordered_set<std::string>> &, int,
                 const std::function<std::pair<std::vector<double>,
                                               std::vector<double>>(
                     int rounds, int warmups,
                     const std::tuple<
                         const Ref<Target> &, const Ref<Device> &,
                         const std::vector<Ref<Array>> &,
                         const std::unordered_map<std::string, Ref<Array>> &>,
                     const std::vector<Func> &funcs)> &>(),
             "schedule"_a, "target"_a, "device"_a, "predict_func"_a,
             "update_func"_a, "tag"_a = "", "min_block_size"_a = 0,
             "random_seed"_a = std::nullopt, "rule_set"_a = std::nullopt,
             "verbose"_a = 0,
             "remote_measure_submit"_a = std::function<
                 std::pair<std::vector<double>, std::vector<double>>(
                     int rounds, int warmups,
                     const std::tuple<const Ref<Target> &, const Ref<Device> &,
                                      const std::vector<Ref<Array>> &,
                                      const std::unordered_map<
                                          std::string, Ref<Array>> &> &att,
                     const std::vector<Func> &funcs)>{})
        .def("set_params", &AutoSchedule::setParams, "args"_a,
             "kws"_a = std::unordered_map<std::string, Ref<Array>>())
        .def("search_one_round", &AutoSchedule::searchOneRound, "n"_a,
             "n_exploit"_a, "n_explore"_a)
        .def("gen_features", &AutoSchedule::genFeatures, "schedules"_a)
        .def("test_and_add", &AutoSchedule::testAndAdd, "sketches"_a)
        .def("get_best_schedule", &AutoSchedule::getBestSchedule)
        .def("test_round", &AutoSchedule::testRound,
             "nth_sketch"_a = std::unordered_map<std::string, int>())
        .def("get_flop", &AutoSchedule::getFlop)
        .def("get_tag", &AutoSchedule::getTag)
        .def("get_best_time", &AutoSchedule::getBestTime);
    m.def("rpc_measure", &rpcMeasure, "rounds"_a, "warmups"_a, "att"_a,
          "funcs"_a);
}

} // namespace freetensor
